src/messages/direct.rs

```diff
@@ -0,0 +1,234 @@
+                 {:?} }}",
+                part_index + 1,
+                part_count,
+                proof.len(),
+                leading_zero_bytes
+            ),
+            ResourceProofResponseReceipt => write!(formatter, "ResourceProofResponseReceipt"),
+            ProxyRateLimitExceeded { ref ack } => {
+                write!(formatter, "ProxyRateLimitExceeded({:?})", ack)
+            }
+            ParsecRequest(ref v, _) => write!(formatter, "ParsecRequest({}, _)", v),
+            ParsecResponse(ref v, _) => write!(formatter, "ParsecResponse({}, _)", v),
+            ParsecPoke(ref v) => write!(formatter, "ParsecPoke({})", v),
+        }
+    }
+}
+
+#[cfg_attr(feature = "mock_serialise", derive(Clone))]
+#[derive(Serialize, Deserialize)]
+pub struct SignedDirectMessage {
+    content: DirectMessage,
+    src_id: PublicId,
+    signature: Signature,
+}
+
+impl SignedDirectMessage {
+    /// Create new `DirectMessage` with `content` and signed by `src_full_id`.
+    pub fn new(content: DirectMessage, src_full_id: &FullId) -> Result<Self, RoutingError> {
+        let signature = self::implementation::sign(src_full_id, &content)?;
+
+        Ok(Self {
+            content,
+            src_id: *src_full_id.public_id(),
+            signature,
+        })
+    }
+
+    /// Verify the message signature.
+    pub fn verify(&self) -> Result<(), RoutingError> {
+        self::implementation::verify(&self.src_id, &self.signature, &self.content)
+    }
+
+    /// The priority Crust should send this message with.
+    pub fn priority(&self) -> u8 {
+        match self.content {
+            DirectMessage::ResourceProofResponse { .. } => 9,
+            _ => 0,
+        }
+    }
+
+    /// Verify the message signature and return its content and the sender id.
+    /// Consume the message in the process.
+    pub fn open(self) -> Result<(DirectMessage, PublicId), RoutingError> {
+        self.verify()?;
+        Ok((self.content, self.src_id))
+    }
+
+    /// Content of the message.
+    #[cfg(any(test, feature = "mock_serialise"))]
+    pub fn content(&self) -> &DirectMessage {
+        &self.content
+    }
+}
+
+impl Debug for SignedDirectMessage {
+    fn fmt(&self, formatter: &mut Formatter) -> fmt::Result {
+        write!(
+            formatter,
+            "SignedDirectMessage {{ content: {:?}, src_id: {:?}, signature: {:?} }}",
+            self.content, self.src_id, self.signature
+        )
+    }
+}
+
+#[cfg(not(feature = "mock_serialise"))]
+mod implementation {
+    use super::*;
+    use maidsafe_utilities::serialisation;
+
+    pub fn sign(src_full_id: &FullId, content: &DirectMessage) -> Result<Signature, RoutingError> {
+        let mut serialised = serialisation::serialise(src_full_id.public_id())?;
+        serialised.extend_from_slice(&serialisation::serialise(content)?);
+
+        let signature = src_full_id.signing_private_key().sign_detached(&serialised);
+        Ok(signature)
+    }
+
+    pub fn verify(
```

Comment:
if you rename `sign`, this should probably be verbosified too for consistency :smile: